### PR TITLE
update to latest gflownet changes

### PR DESCRIPTION
**Pull request by ginihumer**

### CODE REVIEW

File: activelearning/dataset/grid.py [hunk 1]
```diff
@@ -48,7 +48,7 @@ def append(self, X, y):
         append new instances to the data
         """
         # X, y = self.deprocess(X, y) # append data in raw form (i.e., not normalized)
-        X *= self.grid_size
+        X = (X + 1) * self.grid_size / 2  # denormalize from the range [-1; 1]
```
Comment (alexhernandezgarcia) [inline activelearning/dataset/grid.py]: I suspect this will not work as expected in general (for example for both Branin and Hartmann). Can you elaborate what the intention is here?

File: activelearning/dataset/grid.py [hunk 2]
```diff
@@ -57,7 +57,7 @@ def preprocess(self, X, y):
         - normalizes the scores
         - normalizes the states
         """
-        states = X / self.grid_size
+        states = X / self.grid_size * 2 - 1  # normalize to [-1; 1]
```
Comment (alexhernandezgarcia) [inline activelearning/dataset/grid.py]: Same as above

File: activelearning/dataset/grid.py [hunk 3]
```diff
@@ -300,7 +300,11 @@ def get_candidate_set(self, as_dataloader=False):
                 candidate_set,
                 batch_size=self.batch_size,
             )
-        return candidate_set, xi / self.grid_size, yi / self.grid_size
+        return (
+            candidate_set,
+            xi / self.grid_size * 2 - 1,
+            yi / self.grid_size * 2 - 1,
+        )  # scale to [-1; 2]
```
Comment (alexhernandezgarcia) [inline activelearning/dataset/grid.py]: I am also confused by this

File: activelearning/dataset/ocp.py [hunk 4]
```diff
@@ -74,7 +74,7 @@ def __getitem__(self, key):
 
         x, y = self.preprocess(x.to(self.float).squeeze(), y.to(self.float).squeeze())
         if self.return_target:
-            return x, y
+            return x, y * -1  # turn into maximization problem
```
Comment (alexhernandezgarcia) [inline activelearning/dataset/ocp.py:88]: I would document this a bit more clearly. For example at the head of the script.

File: activelearning/oracle/oracle.py [hunk 5]
```diff
@@ -44,10 +45,11 @@ def __init__(
                 do_domain_map=do_domain_map,
                 device=self.device,
                 float_precision=self.float_precision,
+                negate=negate,
             )
 
         def __call__(self, states):
-            return Branin.__call__(self, states.clone())
+            return Branin.proxy2reward(self, Branin.__call__(self, states.clone()))
```
Comment (alexhernandezgarcia) [inline activelearning/oracle/oracle.py]: Would this simpler version not do the same?
```suggestion
            return Branin.rewards(states.clone())
```

Comment (alexhernandezgarcia) [reply activelearning/oracle/oracle.py]: I also think clone is not needed any more, but I'm not sure.

File: activelearning/oracle/oracle.py [hunk 6]
```diff
@@ -69,10 +78,11 @@ def __init__(self, cost=1, fidelity=1, device="cpu", float_precision=64):
                 fidelity=fidelity,
                 device=self.device,
                 float_precision=self.float_precision,
+                negate=negate,
             )
 
         def __call__(self, states):
-            return Hartmann.__call__(self, states.clone())
+            return Hartmann.proxy2reward(self, Hartmann.__call__(self, states.clone()))
```
Comment (alexhernandezgarcia) [inline activelearning/oracle/oracle.py:85]: As with Branin:
```suggestion
            return Hartmann.rewards(states.clone())
```

**Review by alexhernandezgarcia — COMMENTED** [review]
Looks good but I have left some important questions about the denormalisation of the data.

File: activelearning/dataset/grid.py (hunk 1 above)
Comment (ginihumer) [reply activelearning/dataset/grid.py]: without normalization we would have grid values in the range of [0; grid_size] to scale between [0; 1] we would just divide by grid_size. to scale between [-1; 1] (as we need to in the new version) we can divide by half the grid size and subtract 1. 

if we want to add new selected datapoints (which are selected between [-1; 1]), and add it to the base data that is not normalized, we also have to denormalize the new data to be in the original range of [0; grid_size], which is the inverse of the above (add 1 and multiply by half the grid_size)

does that make sense? why would you assume that it does not work like this?

File: activelearning/dataset/grid.py (hunk 2 above)
Comment (ginihumer) [reply activelearning/dataset/grid.py]: what exactly do you think is the problem with this?
do you think that the normalization before was ok?

File: activelearning/dataset/grid.py (hunk 3 above)
Comment (ginihumer) [reply activelearning/dataset/grid.py]: whoops, should be [-1; 1]
same normalization as before

File: activelearning/oracle/oracle.py (hunk 5 above)
Comment (ginihumer) [reply activelearning/oracle/oracle.py]: this will trigger an endless loop because it calls the __call__ function, which will enter the BraninOracle's call function again

File: activelearning/oracle/oracle.py (hunk 6 above)
Comment (ginihumer) [reply activelearning/oracle/oracle.py:85]: same

File: activelearning/dataset/grid.py (hunk 1 above)
Comment (alexhernandezgarcia) [reply activelearning/dataset/grid.py]: Ok, this is going to need some additional documentation because even I am confused. And in the future, we will have data from other environments and it is very important to know in which format the input data is and in which format the outputs should be.

Comment (alexhernandezgarcia) [reply activelearning/dataset/grid.py]: Also, unclear what "raw" means. In the GFlowNet repo, we have these formats for states/data:
- "Environment format": the representation of the states in a GFlowNet environment. For example, for the grid, that is `[0, length]`.
- Proxy format: the representation received by a Proxy, or an Oracle in the active learning sense. That's the output of `Env.states2proxy(states)`.
- Policy format: the representation of the states passed to the policy models.
- Readable: a string representation meant to be more "human-readable"

Comment (alexhernandezgarcia) [reply activelearning/dataset/grid.py]: I would use these terms if possible.

And more importantly, I would (re-)use the methods from the GFlowNet environment to do the conversions.

Comment (alexhernandezgarcia) [reply activelearning/dataset/grid.py]: So, this said, let's try to clarify:
- In what format are the inputs?
- In what format should the outputs be?

File: activelearning/dataset/grid.py (hunk 2 above)
Comment (alexhernandezgarcia) [reply activelearning/dataset/grid.py]: See my comments on the other thread. I think it is very important here to write a good docstring.

File: activelearning/dataset/grid.py (hunk 1 above)
Comment (ginihumer) [reply activelearning/dataset/grid.py]: - the inputs are in the range of [0; grid_size] 
- the outputs in the range [-1; 1]

I am not sure if we should use the term "proxy" at all in this context... in this active learning repo we only have the components "surrogate", "acquisition", "sampler", "selector", and "oracle". I think we should keep the terms more generic and not specific to the GFN task. what do you think?

Comment (ginihumer) [reply activelearning/dataset/grid.py]: or do you want to generally introduce proxies into the active learning workflow? currently it is only a thing when using GFNs as sampler

Comment (alexhernandezgarcia) [reply activelearning/dataset/grid.py]: Fine with me, but it's important to keep in mind that Oracle = GFN Proxy, if we use GFN. We can use the term "Oracle format" if you prefer. Then, we have the following:
- Inputs are in "environment format"
- Outputs are in "oracle format"
Do you agree? If so, I would document that. And like I said above I would reuse the method `states2proxy()`.

Comment (alexhernandezgarcia) [reply activelearning/dataset/grid.py]: > or do you want to generally introduce proxies into the active learning workflow? currently it is only a thing when using GFNs as sampler

Well, not only. When you use Branin (for example), you are using a GFN proxy, even if you don't use GFN as the sampler. See here: https://github.com/alexhernandezgarcia/activelearning/blob/ch/updateGFN/activelearning/oracle/oracle.py#L24

Comment (ginihumer) [reply activelearning/dataset/grid.py]: hmm... right

I'll have to think about how to best structure it, but ideally with a clear head and nose^^

File: activelearning/oracle/oracle.py (hunk 5 above)
Comment (alexhernandezgarcia) [reply activelearning/oracle/oracle.py]: Uhm, ok. There must be a way to do this in a simpler way, but it's fine, not a deal breaker.

File: activelearning/dataset/grid.py [hunk 7]
```diff
@@ -22,12 +23,14 @@ def __init__(
         X_data,
         y_data=None,
         normalize_scores=True,
+        state2result=None,
```
Comment (alexhernandezgarcia) [inline activelearning/dataset/grid.py]: I think it's important to add docstring and type hints here. That will be the way to be sure about the format/representations that are expected.

File: activelearning/dataset/grid.py [hunk 8]
```diff
@@ -22,12 +23,14 @@ def __init__(
...
         float=torch.float64,
     ):
         super().__init__(X_data, y_data, float=float)
         self.normalize_scores = normalize_scores
         self.grid_size = grid_size
         self.stats = self.get_statistics(y_data)
+        self.state2result = state2result
```
Comment (alexhernandezgarcia) [inline activelearning/dataset/grid.py]: What does result mean in this context? Explain in docstring? :sweat_smile:

File: activelearning/dataset/grid.py [hunk 9]
```diff
@@ -97,44 +91,24 @@ def __init__(
...
-        self.normalize_scores = normalize_scores
         self.train_fraction = train_fraction
         self.train_path = train_path
         self.test_path = test_path
-        self.grid_size = grid_size
 
         self.initialise_dataset()
 
-    def statetorch2readable(self, state, alphabet={}):
+    def proxy2state(self, proxy_state):
```
Comment (alexhernandezgarcia) [inline activelearning/dataset/grid.py]: Why/where is this method needed? I'm asking because if we can do things without the need for these method, it would be great. One, this isn't needed in the gflownet repo and hence is not implemented. So it would have to be implemented for each new environment here (not great). Second, in some cases, there will not be a one-to-one mapping from proxy to state.

**Review by alexhernandezgarcia — CHANGES_REQUESTED** [review]
Looks good in general! 

As I wrote in some comments, I think it's important to add documentation and type hints in some places. Other than that, I have dropped a couple of questions, but that's it.

File: activelearning/dataset/grid.py (hunk 9 above)
Comment (ginihumer) [reply activelearning/dataset/grid.py]: we need it because the sampled results are in proxy format and must be mapped back to environment format for storing. this is not needed in gflownets generally because we don't add any new datapoints to the train set

Comment (alexhernandezgarcia) [reply activelearning/dataset/grid.py]: Uhm, but if you sample candidates from a GFlowNet, they will/can be in environment format.

File: activelearning/dataset/grid.py [hunk 10]
```diff
@@ -13,6 +15,8 @@ class GridData(Data):
         y_data: array(N, 1); scores at each position
         normalize_scores: bool; maps the scores to [0; 1]
         grid_size: int; specifies the width and height of the grid (grid_size x grid_size); used to normalize the state positions
+        state2result: function that takes raw states (X_data) and transforms them into the desired format;
```
Comment (alexhernandezgarcia) [inline activelearning/dataset/grid.py]: What format is "raw states"? (sorry I'm being picky here but it's for a good reason). Confusion with data representations has been the source of a lot of pain and wasted time and I would like to avoid this in the future.

File: activelearning/dataset/grid.py (hunk 9 above)
Comment (ginihumer) [reply activelearning/dataset/grid.py]: interesting! That is the better option then. I just have to figure out how to adopt this for the baseline samplers

Comment (alexhernandezgarcia) [reply activelearning/dataset/grid.py]: So to clarify: the environment format is the baseline (main) format, from which the others (proxy, policy, readable) can be obtained. So that one is always available. That's why I would suggest that storage is done in environment format.

File: tests/dataset/test_ocp.py [hunk 11]
```diff
@@ -0,0 +1,47 @@
+import pytest
+import torch
+from activelearning.dataset.ocp import OCPDatasetHandler
+
+
+@pytest.fixture
+def ocp_dataset_handler_base():
+    ocp_checkpoint_path = "/network/scratch/a/alexandre.duval/ocp/runs/4648581/checkpoints/best_checkpoint.pt"
+    dataset_path = "/network/scratch/a/alexandre.duval/ocp/runs/4657270/deup_dataset"
```
Comment (alexhernandezgarcia) [inline tests/dataset/test_ocp.py:10]: This will crash for most people because the files/dir will not exist. If you want to leave this (I'm not against it), I would at least surround it with a `try: ... except: ...` so that that the test does not fail because of that.

File: tests/dataset/test_ocp.py [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+    dataset_handler = OCPDatasetHandler(
+        ocp_checkpoint_path, dataset_path, float_precision=32
+    )
+    return dataset_handler
+
+
+@pytest.fixture
+def ocp_dataset_handler_train_split():
+    ocp_checkpoint_path = "/network/scratch/a/alexandre.duval/ocp/runs/4648581/checkpoints/best_checkpoint.pt"
+    dataset_path = "/network/scratch/a/alexandre.duval/ocp/runs/4657270/deup_dataset"
```
Comment (alexhernandezgarcia) [inline tests/dataset/test_ocp.py:29]: Same as above

**Review by alexhernandezgarcia — APPROVED** [review]
The most important bit are my comments about the OCP tests, that will crash for most other users.

Once that is done, you may go ahead and merge.

I wrote it somewhere in the comments, but in my opinion the new term "raw states" is likely to cause more confusion than not, since it's a new term that refers to something that already exists (simply "states", which are the states in "environment format". I think it's easier to just use "states" to refer to the standard states, and specify when they are converted into something else. I'm saying this from a position of having thought a lot about this for the gflownet repo, and discuss solutions with other contributors, etc.. Anyway, not a deal breaker for me.
